### PR TITLE
[opentitantool] Add interactive console

### PR DESCRIPTION
**Pull request by cfrantz**
Add an interactive `console` command.  The console command includes
options which allow it to be useful by automation as well:

`--logfile=<file>` logs the console output to a file.
`--timeout=<seconds>` causes the console to exit after `seconds`.
`--exit_success=<regex>` causes the console to exit successfully after
observing `regex`.
`--exit_success=<regex>` causes the console to exit with failure after
observing `regex`.

Signed-off-by: Chris Frantz <cfrantz@google.com>

### CODE REVIEW

**Comment by cfrantz** [comment]
This PR builds on #7362.  You need only review the "interactive console" change in this PR.

File: sw/host/opentitantool/src/command/console.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+impl InnerConsole {
+    const CTRL_C: u8 = 3;
+    const BUFFER_LEN: usize = 1024;
+
+    // Runs an interactive console until CTRL_C is received.
+    fn interact<R>(
+        &mut self,
+        transport: &mut dyn Transport,
+        stdin: &mut R,
```
Comment (mcy) [inline sw/host/opentitantool/src/command/console.rs:117]: You can write `&mut impl (Read + AsRawFd)`.

File: sw/host/opentitantool/src/command/console.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                    // should be an error.
+                    if self.exit_success.is_some() {
+                        return Err(anyhow!("Console timeout exceeded"));
+                    } else {
+                        break;
+                    }
+                }
+            }
+
+            // This loop _should_ really use `poll` to learn when the
```
Comment (mcy) [inline sw/host/opentitantool/src/command/console.rs:139]: Maybe a thing to do in the future is to add a poll() function that rudely blocks when you can't find the fd?

File: sw/host/opentitantool/src/command/console.rs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+                ExitStatus::ExitSuccess => {
+                    break;
+                }
+                ExitStatus::ExitFailure => {
+                    return Err(anyhow!("Matched exit_failure expression"));
+                }
+            };
+
+            // Wait for input from the user.
+            match file::wait_read_timeout(&*stdin, Duration::from_millis(0)) {
```
Comment (mcy) [inline sw/host/opentitantool/src/command/console.rs:159]: Why not just

if wait_read_timeout().is_ok() {
 ...
}

File: sw/host/opentitantool/src/command/console.rs (hunk 1 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/console.rs:117]: Done.

File: sw/host/opentitantool/src/command/console.rs (hunk 3 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/console.rs:159]: Done.

File: sw/host/opentitantool/src/command/console.rs (hunk 2 above)
Comment (cfrantz) [reply sw/host/opentitantool/src/command/console.rs:139]: The `Uart` trait's `read_timeout` does exactly what you suggest: blocks on that single resource for a user-specified period of time.  The comment is lamenting the fact that the "proper" way to do things is via the unix poll (and/or friends) syscall to block on a set of resources and wake up when one of them becomes ready.

I've added a TODO to re-evaluate this once we have more backends written and can make a better decision about multi-fd blocking.

File: sw/host/opentitantool/src/command/console.rs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                }
+
+                // If we have exit condition regexes check them.
+                self.append_buffer(&buf[..len]);
+                if self
+                    .exit_success
+                    .as_ref()
+                    .map(|rx| rx.is_match(&self.buffer))
+                    == Some(true)
+                {
```
Comment (mcy) [inline sw/host/opentitantool/src/command/console.rs:206]: This is the sensible way to do it but omg it's so ridiculous.

**Review by mcy — APPROVED** [review]

**Pull request merged**